public/app/plugins/datasource/loki/querybuilder/components/NestedQuery.test.tsx

```diff
@@ -0,0 +1,112 @@
+import { fireEvent, render, screen, waitFor } from '@testing-library/react';
+import userEvent from '@testing-library/user-event';
+import React from 'react';
+
+import { DataSourceInstanceSettings, DataSourcePluginMeta } from '@grafana/data';
+
+import { LokiDatasource } from '../../datasource';
+import { LokiVisualQueryBinary } from '../types';
+
+import { NestedQuery, Props as NestedQueryProps } from './NestedQuery';
+
+type Operator = '+' | '-' | '*' | '/' | '%' | '^' | '==' | '!=' | '>' | '<' | '>=' | '<=';
+type VectorMatchType = 'on' | 'ignoring';
+
+const createMockProps = (
+  operator: Operator = '/',
+  vectorMatchesType: VectorMatchType = 'on',
+  showExplain = false
+): NestedQueryProps => {
+  const nestedQuery: LokiVisualQueryBinary = {
+    operator: operator,
+    vectorMatchesType: vectorMatchesType,
+    query: {
+      labels: [],
+      operations: [],
+    },
+  };
+
+  const datasource = new LokiDatasource(
```

Comment:
You can use this function for this https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/mocks.ts#L32

The advantage is isolating mock datasource generation into a single place, with a common method.

Reply:
Show me the error. It's possible that another mock method should be added here: https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/mocks.ts#L29

Currently there's no recommended or "good" way of mocking `templateSrv`.

Reply:
@matyax nvm, it works just using `const datasource = createLokiDatasource();`
i think i just tried overcomplicating it by adding args but the defaults work fine.